Protect: Add Home page

dkmyta opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->
<!-- Would you like this feature to be tested by Beta testers?
Please add testing instructions to projects/plugins/jetpack/to-test.md in a new commit as part of your PR. -->
<!-- a12s: If you have an expected version that you're aiming for the PR to add, please use the Milestone field to communicate it. If you leave it blank, that indicates there isn't a preference. -->

## Description
Add Home tab, page and stat cards.

### Other information:

- [ ] Have you written new tests for your changes, if applicable?
- [ ] Have you checked the E2E test CI results, and verified that your changes do not break them?
- [ ] Have you tested your changes on WordPress.com, if applicable (if so, you'll see a generated comment below with a script to run)?

## Jetpack product discussion
<!-- If you're an Automattician, include a shortlink to the p2 discussion with Jetpack Product here. -->
<!-- Make sure any changes to existing products have been discussed and agreed upon -->
- https://github.com/Automattic/jetpack-scan-team/issues/1565

## Does this pull request change what data or activity we track or use?
<!--- If so, please add the "[Status] Needs Privacy Updates" label and explain what changes there are. -->
<!--- Check existing Jetpack support documents for a preview of the information we need. -->

## Testing instructions:
<!-- If you were reviewing this PR, how would you like the instructions to be presented? -->
<!-- Please include detailed testing steps, explaining how to test your change. -->
<!-- Bear in mind that context you working on is not obvious for everyone.  -->
<!-- Adding "simple" configuration steps will help reviewers to get to your PR as quickly as possible. -->
<!-- "Before / After" screenshots can also be very helpful when the change is visual. -->

* Checkout this branch load Protect in JT
* Ensure you are redirected to the Home tab after passing setup
* Review the various states of the HomeStatCards error, scanning, idle, disabled
* Ensure no regressions are introduced

## Screenshots
<img width="1639" alt="Screen Shot 2024-12-03 at 10 51 06" src="https://github.com/user-attachments/assets/03241100-fa36-4122-a7a4-fa94748066c1">

nateweller committed:
Init project branch

nateweller committed:
Init project branch

nateweller committed:
Protect: add ShieldIcon component

Protect: Add ShieldIcon Component (#40402)

Components: Add ScanReport (#40419)

dkmyta commented on projects/plugins/protect/src/class-jetpack-protect.php:
```diff
@@ -457,8 +457,9 @@ public static function get_waf_stats() {
 
 		return array(
-			'blockedRequests'           => Plan::has_required_plan() ? Waf_Stats::get_blocked_requests() : false,
+			'blockedRequests'           => Waf_Stats::get_blocked_requests(),
```
Makes stats available regardless of plan

nateweller commented:
<img width="1241" alt="Screenshot 2024-12-05 at 4 08 03 PM" src="https://github.com/user-attachments/assets/4a8ea53a-8261-474d-bf34-11127066ee23">

- [ ] Header alignment issue when scan in progress on a free plan

nateweller commented:
- [ ] Scan upgrade prompt should redirect to `/scan` after checkout

dkmyta commented:
>Header alignment issue when scan in progress on a free plan

Darnit, I had this fixed previously - I think a few things unforuntately got overwritten in the most recent rebase :(

nateweller commented:
- [ ] Mismatching line counts across stat cards, we could either apply `whitespace: nowrap` here as has been done in the Firewall tab header, or update the labels to all wrap two lines?

<img width="634" alt="Screenshot 2024-12-05 at 4 18 27 PM" src="https://github.com/user-attachments/assets/77415970-e2d1-4491-bf64-edc3d25d3589">

dkmyta commented:
>Mismatching line counts across stat cards, we could either apply whitespace: nowrap here as has been done in the Firewall tab header, or update the labels to all wrap two lines?

This was a discussion with @ilonagl - the issue with `nowrap` was that when we are viewing free results the scan card uses the term `Vulnerabilities` which is too long to fit on a single line. The solution was wrapping the text, but I agree with ensuring there a two lines always 👍🏻

dkmyta commented:
>Scan upgrade prompt should redirect to /scan after checkout

I suppose we'll need to upgrade all applicable uses of `upgradePlan` to supply a redirect now. I wonder if there is also an opportunity here to add a upgrade CTA to the new Home page 🤔

nateweller commented:
<img width="1585" alt="Screenshot 2024-12-06 at 11 23 32 AM" src="https://github.com/user-attachments/assets/49fad2ac-ebaf-4754-bafe-e57a4e3d7914">

<img width="1308" alt="Screenshot 2024-12-06 at 11 25 16 AM" src="https://github.com/user-attachments/assets/fd296daa-7477-4c5d-95ae-808fe41ab4af">

- [ ] The stat cards start wrapping into the next line on < 1400px screen widths, and overflowing horizontally when they are lined up.

nateweller commented on projects/plugins/protect/src/js/routes/home/home-statcards.jsx:
```diff
@@ -0,0 +1,243 @@
+	const variant = isSmall ? 'horizontal' : 'square';
+
+	const { data: status } = useScanStatusQuery();
+	const scanning = isScanInProgress( status );
```
This should be in a `useMemo`

dkmyta commented:
>The stat cards start wrapping into the next line on < 1400px screen widths, and overflowing horizontally when they are lined up.

Chatted with @ilonagl about how we should go about handling this, obviously not ideal that it wrap at that size as its a pretty common view port size - also not ideal to have it overflow. Open to suggestions!

nateweller commented on projects/plugins/protect/src/js/routes/home/home-statcards.jsx:
```diff
@@ -0,0 +1,243 @@
+
+	const iconHeight = useMemo( () => 20, [] );
+
+	const lastCheckedMessage = useMemo( () => {
```
We will likely need to consider prioritizing the "scanning" state above the "error" state, for when we are optimistically scanning/expecting a scan to start soon. Similar to how we decide when to show the scan in progress header ( is scanning -> is error -> results ).

Otherwise, I am seeing the following on the first load of the home page on a new site:

<img width="1280" alt="Screenshot 2024-12-06 at 11 22 32 AM" src="https://github.com/user-attachments/assets/5f0f8fb5-117d-418e-a6af-cb8b9bf71709">

nateweller commented on projects/plugins/protect/src/js/routes/home/home-statcards.jsx:
```diff
@@ -0,0 +1,243 @@
+			return __( 'Your results will be ready soon.', 'jetpack-protect' );
+		}
+
+		const entityLabel = hasPlan
+			? _n( 'threat', 'threats', numThreats, 'jetpack-protect' )
+			: _n( 'vulnerability', 'vulnerabilities', numThreats, 'jetpack-protect' );
+
+		if ( lastCheckedLocalTimestamp ) {
+			if ( numThreats > 0 ) {
+				return sprintf(
+					// translators: %1$s: date/time, %2$d: number, %3$s: entity label
+					__( 'Last checked on %1$s: We found %2$d %3$s.', 'jetpack-protect' ),
+					lastCheckedLocalTimestamp,
+					numThreats,
+					entityLabel
+				);
+			}
+			return sprintf(
+				// translators: %s: date/time
+				__( 'Last checked on %s: Your site is secure.', 'jetpack-protect' ),
+				lastCheckedLocalTimestamp
+			);
+		}
+
+		return sprintf(
+			// translators: %1$s: date/time, %2$d: number, %3$s: entity label
+			__( 'Last scan we found %2$d %3$s.', 'jetpack-protect' ),
+			lastCheckedLocalTimestamp,
+			numThreats,
+			entityLabel
+		);
```
We should avoid concatenating translated strings together.

See FG: PCYsg-3hI-p2#it-s-ok-to-almost-repeat-yourself

To avoid this, we will need to repeat some of the strings, i.e. `hasPlan ? __( 'We identified %s threats' ) : __( 'We identified %s vulnerabilities' )`

I realize we do this in a few places already in Protect, will need to address those existing instances as we come across them as well :+1:

dkmyta replied:
Thanks for the link, good to know! I have updated to avoid that approach in [5728962](https://github.com/Automattic/jetpack/pull/40317/commits/57289626b75a733d1cd92c7ff0543b97f725f823) - its a little more complicated now since we needed to consider `hasPlan` for `threat` vs `vulnerability` _and_ `numThreats` for plurals. Will keep this all in mind moving forward.

nateweller commented on projects/plugins/protect/src/js/routes/home/home-statcards.jsx:
```diff
@@ -0,0 +1,243 @@
+		blockedLogins: allTimeBlockedLoginsCount = 0,
+	} = stats || {};
+
+	const iconHeight = useMemo( () => 20, [] );
```
```suggestion
	const ICON_HEIGHT = 20;
```
Don't need to memo a constant value.

Nit would be using a const-style naming convention like `ICON_HEIGHT`. I would even pull it out of inside the component so that it only is declared/initialized once.

nateweller commented on projects/plugins/protect/src/js/routes/home/home-statcards.jsx:
```diff
@@ -0,0 +1,243 @@
+	const [ isSmall ] = useBreakpointMatch( [ 'sm', 'lg' ], [ null, '<' ] );
+	const variant = isSmall ? 'horizontal' : 'square';
+
+	const { data: status } = useScanStatusQuery();
```
Just thinking/wondering out loud, could we enable `usePolling` for the scan status query and have "live" scanning data in this card? Maybe one day!

dkmyta replied:
I imagine in the main `HomePage` component we'll want `usePolling` enabled so the card has the most recent data as the scan run - I wonder if we could pass status data down from there instead to ensure things remain up to date?

nateweller commented:
> > The stat cards start wrapping into the next line on < 1400px screen widths, and overflowing horizontally when they are lined up.
> 
> Chatted with @ilonagl about how we should go about handling this, obviously not ideal that it wraps at that size as its a pretty common view port size - also not ideal to have it overflow. Open to suggestions!

There's plenty of room, I would allow the stat cards to hold their natural width and have the "your site is safe with us" message fit/wrap:

<img width="1068" alt="Screenshot 2024-12-06 at 11 57 20 AM" src="https://github.com/user-attachments/assets/4284d72d-9b47-43ff-b8f7-536d99704a8d">

nateweller reviewed (changes requested):
Looks great, works well, couple notes remaining!

dkmyta commented:
>There's plenty of room, I would allow the stat cards to hold their natural width and have the "your site is safe with us" message fit/wrap:

@nateweller I am not finding the proper combination of styles that would make this fit as you've described, it just seems to overflow. What adjustments did you apply to what elements to make that happen?

Commit:
Protect: add home page

Rebase on admin hero section refactor

Revert accidental type changes

Remove testing code

Init project branch

Protect: Add ShieldIcon Component (#40402)

Refactor AdminSectionHero component

nateweller committed:
changelog

nateweller committed:
Fix translated string concat
